File: ios-oc/VoiceAgent/Chat/ChatSessionView.m
//
//  ChatSessionView.m
//  VoiceAgent
//
//  Created by qinhui on 2025/11/17.
//

#import "ChatSessionView.h"
#import "TranscriptMessageCell.h"
#import <Masonry/Masonry.h>

static inline UIColor *VAHexColor(NSUInteger hexValue, CGFloat alpha) {
    return [UIColor colorWithRed:((hexValue >> 16) & 0xFF) / 255.0
                           green:((hexValue >> 8) & 0xFF) / 255.0
                            blue:(hexValue & 0xFF) / 255.0
                           alpha:alpha];
}

@interface ChatSessionView ()

@property (nonatomic, strong, readwrite) UITableView *tableView;
@property (nonatomic, strong, readwrite) AgentStateView *statusView;
@property (nonatomic, strong) UIView *controlBarView;
@property (nonatomic, strong, readwrite) UIButton *micButton;
@property (nonatomic, strong, readwrite) UIButton *endCallButton;

@end

@implementation ChatSessionView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setupUI];
        [self setupConstraints];
    }
    return self;
}

- (instancetype)init {
    return [self initWithFrame:CGRectZero];
}

- (void)setupUI {
    self.backgroundColor = [UIColor clearColor];
    
    // TableView for transcripts
    self.tableView = [[UITableView alloc] init];
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    self.tableView.backgroundColor = VAHexColor(0x1E293B, 0.5);
    self.tableView.layer.cornerRadius = 12;
    self.tableView.layer.borderWidth = 0.5;
    self.tableView.layer.borderColor = VAHexColor(0x334155, 0.5).CGColor;
    self.tableView.clipsToBounds = YES;
    [self.tableView registerClass:[TranscriptMessageCell class] forCellReuseIdentifier:@"TranscriptMessageCell"];
    [self addSubview:self.tableView];
    
    // Status View
    self.statusView = [[AgentStateView alloc] init];
    [self addSubview:self.statusView];
    
    // Control Bar
    self.controlBarView = [[UIView alloc] init];
    self.controlBarView.backgroundColor = VAHexColor(0x1E293B, 0.8);
    self.controlBarView.layer.cornerRadius = 12;
    self.controlBarView.layer.borderWidth = 0.5;
    self.controlBarView.layer.borderColor = VAHexColor(0x334155, 0.5).CGColor;
    [self addSubview:self.controlBarView];
    
    // Mic Button
    self.micButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.micButton setImage:[UIImage systemImageNamed:@"mic.fill"] forState:UIControlStateNormal];
    self.micButton.tintColor = VAHexColor(0xCBD5E1, 1.0);
    self.micButton.backgroundColor = VAHexColor(0x334155, 1.0);
    self.micButton.layer.cornerRadius = 22;
    [self.controlBarView addSubview:self.micButton];
    
    // End Call Button
    self.endCallButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.endCallButton setTitle:@"Stop Agent" forState:UIControlStateNormal];
    [self.endCallButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    self.endCallButton.titleLabel.font = [UIFont systemFontOfSize:14 weight:UIFontWeightMedium];
    self.endCallButton.backgroundColor = VAHexColor(0xDC2626, 1.0);
    self.endCallButton.layer.cornerRadius = 8;
    [self.controlBarView addSubview:self.endCallButton];
}

- (void)setupConstraints {
    [self.tableView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self);
        make.left.right.equalTo(self).inset(20);
        make.bottom.equalTo(self.statusView.mas_top).offset(-8);
    }];
    
    [self.statusView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.equalTo(self).inset(20);
        make.bottom.equalTo(self.controlBarView.mas_top).offset(-8);
        make.height.mas_equalTo(40);
    }];
    
    [self.controlBarView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.equalTo(self).inset(20);
        make.bottom.equalTo(self.mas_safeAreaLayoutGuideBottom).offset(-20);
        make.height.mas_equalTo(60);
    }];
    
    [self.endCallButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self.controlBarView);
        make.right.equalTo(self.controlBarView).offset(-12);
        make.height.mas_equalTo(36);
        make.width.mas_equalTo(120);
    }];

    [self.micButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self.controlBarView);
        make.right.equalTo(self.endCallButton.mas_left).offset(-12);
        make.width.height.mas_equalTo(44);
    }];
}

- (void)updateMicButtonState:(BOOL)isMuted {
    NSString *imageName = isMuted ? @"mic.slash.fill" : @"mic.fill";
    [self.micButton setImage:[UIImage systemImageNamed:imageName] forState:UIControlStateNormal];
    self.micButton.tintColor = isMuted ? VAHexColor(0xF87171, 1.0) : VAHexColor(0xCBD5E1, 1.0);
    self.micButton.backgroundColor = isMuted ? VAHexColor(0xEF4444, 0.2) : VAHexColor(0x334155, 1.0);
}

- (void)updateStatusView:(NSInteger)state {
    [self.statusView updateState:state];
}

- (void)reloadData {
    [self.tableView reloadData];
}

- (void)scrollToBottomAnimated:(BOOL)animated {
    NSInteger rowCount = [self.dataSource numberOfTranscriptsInChatSessionView:self];
    if (rowCount > 0) {
        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:rowCount - 1 inSection:0];
        [self.tableView scrollToRowAtIndexPath:indexPath atScrollPosition:UITableViewScrollPositionBottom animated:animated];
    }
}

// MARK: - UITableViewDataSource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.dataSource numberOfTranscriptsInChatSessionView:self];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    TranscriptMessageCell *cell = [tableView dequeueReusableCellWithIdentifier:@"TranscriptMessageCell" forIndexPath:indexPath];
    Transcript *transcript = [self.dataSource chatSessionView:self transcriptAtIndex:indexPath.row];
    [cell configureWithTranscript:transcript];
    return cell;
}

// MARK: - UITableViewDelegate
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return UITableViewAutomaticDimension;
}

@end

File: ios-swift/VoiceAgent/Chat/ChatSessionView.swift
//
//  ChatBackgroundView.swift
//  VoiceAgent
//
//  Created by qinhui on 2025/11/17.
//

import UIKit
import SnapKit

class ChatSessionView: UIView {
    // MARK: - UI Components
    let tableView = UITableView()
    let statusView = AgentStateView()
    private let controlBarView = UIView()
    let micButton = UIButton(type: .custom)
    let endCallButton = UIButton(type: .custom)

    // MARK: - Initialization
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
        setupConstraints()
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - UI Setup
    private func setupUI() {
        backgroundColor = .clear

        // TableView — card appearance (rounded corners, border, background)
        tableView.separatorStyle = .none
        tableView.backgroundColor = AppColors.bgCard
        tableView.layer.cornerRadius = 12
        tableView.layer.borderWidth = 0.5
        tableView.layer.borderColor = AppColors.borderDefault.cgColor
        tableView.clipsToBounds = true
        tableView.register(TranscriptMessageCell.self, forCellReuseIdentifier: TranscriptMessageCell.reuseIdentifier)
        addSubview(tableView)

        // Agent status — between tableView and control bar (matches Android layout)
        statusView.isHidden = true
        addSubview(statusView)

        // Control Bar — horizontal strip at bottom
        controlBarView.backgroundColor = AppColors.bgControlBar
        controlBarView.layer.cornerRadius = 12
        controlBarView.layer.borderWidth = 0.5
        controlBarView.layer.borderColor = AppColors.borderDefault.cgColor
        addSubview(controlBarView)

        // Mic Button (right side)
        micButton.setImage(UIImage(systemName: "mic.fill")?.withRenderingMode(.alwaysTemplate), for: .normal)
        micButton.tintColor = AppColors.micNormalIcon
        micButton.backgroundColor = AppColors.micNormalBg
        micButton.layer.cornerRadius = 22
        controlBarView.addSubview(micButton)

        // Stop Button — wide pill shape
        endCallButton.setTitle("Stop Agent", for: .normal)
        endCallButton.setTitleColor(.white, for: .normal)
        endCallButton.titleLabel?.font = .systemFont(ofSize: 14, weight: .medium)
        endCallButton.backgroundColor = AppColors.btnStopBg
        endCallButton.layer.cornerRadius = 8
        endCallButton.contentEdgeInsets = UIEdgeInsets(top: 0, left: 16, bottom: 0, right: 16)
        controlBarView.addSubview(endCallButton)
    }

    private func setupConstraints() {
        // Control bar at bottom
        controlBarView.snp.makeConstraints { make in
            make.left.right.equalToSuperview().inset(20)
            make.bottom.equalTo(safeAreaLayoutGuide).offset(-20)
            make.height.equalTo(60)
        }

        // Stop button — right side of control bar
        endCallButton.snp.makeConstraints { make in
            make.right.equalToSuperview().offset(-12)
            make.centerY.equalToSuperview()
            make.height.equalTo(36)
        }

        // Mic button — left of stop button
        micButton.snp.makeConstraints { make in
            make.right.equalTo(endCallButton.snp.left).offset(-12)
            make.centerY.equalToSuperview()
            make.width.height.equalTo(44)
        }

        // Agent status — between tableView and control bar
        statusView.snp.makeConstraints { make in
            make.left.right.equalToSuperview().inset(20)
            make.bottom.equalTo(controlBarView.snp.top).offset(-8)
            make.height.equalTo(40)
        }

        // TableView fills the space above status view
        tableView.snp.makeConstraints { make in
            make.top.equalToSuperview()
            make.left.right.equalToSuperview().inset(20)
            make.bottom.equalTo(statusView.snp.top).offset(-8)
        }
    }

    // MARK: - Public Methods
    func updateMicButtonState(isMuted: Bool) {
        let imageName = isMuted ? "mic.slash.fill" : "mic.fill"
        micButton.setImage(UIImage(systemName: imageName)?.withRenderingMode(.alwaysTemplate), for: .normal)
        micButton.tintColor = isMuted ? AppColors.micMutedIcon : AppColors.micNormalIcon
        micButton.backgroundColor = isMuted ? AppColors.micMutedBg : AppColors.micNormalBg
    }

    func updateStatusView(state: AgentState) {
        statusView.updateState(state)
    }

    // UIKit may reset the table background color to a system dynamic color during layout.
    // Apply it from the parent controller after the table has been attached to the final view hierarchy.
    func applyTableBackgroundWorkaround() {
        tableView.backgroundColor = UIColor(hex: 0x1E293B, alpha: 0.5)
    }
}
